### PR TITLE
[RFC][HMA-in-a-bottle] Possible example of creating a modular matcher

### PR DESCRIPTION
**Pull request by Sam-Freeman**
Summary
---------

An example of decoupling the matcher logic from the CLI into its own module and generic base class which allows for creation of specific matcher 'types' -- this was the simplest way to separate the logic for file matching and raw hash matching. 

Test Plan
---------

This is an RFC, but ran it locally and it all works.

### CODE REVIEW

File: python-threatexchange/threatexchange/cli/match_cmd.py [hunk 1]
```diff
@@ -145,104 +146,41 @@ def __init__(
...
+            for r in results:
+                metadatas: t.List[t.Tuple[str, FetchedSignalMetadata]] = r.metadata
+                for collab, fetched_data in metadatas:
+                    if not self.all and collab in seen:
+                        continue
+                    seen.add(collab)
+                    # Supposed to be without whitespace, but let's make sure
+                    distance_str = "".join(r.similarity_info.pretty_str().split())
+                    print(
+                        # s_type.get_name(),
```
Comment (BarrettOlson) [inline python-threatexchange/threatexchange/cli/match_cmd.py:182]: hmm, knowing the type matched is valuable. One options though I'm not sure it is worth the overhead is to have it as part of the metadata(s)

Comment (Sam-Freeman) [reply python-threatexchange/threatexchange/cli/match_cmd.py:182]: Yeah, agreed, and something that we/I need to come back to. I commented it out for now as it didn't exist in this scope, and didn't want to forget that I removed it.

File: python-threatexchange/threatexchange/matcher/matcher.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from threatexchange.content_type.content_base import ContentType
+from threatexchange.signal_type.index import IndexMatch, SignalTypeIndex
+from threatexchange.signal_type.signal_base import SignalType
+
+T = t.TypeVar("T")
+
+
+class Matcher(t.Generic[T]):
+    # question (sa) -- how do we want to handle the settings? In the CLI it's considered
+    # a god object -- do we want to pass it to the init of the class?
```
Comment (dougneal) [inline python-threatexchange/threatexchange/matcher/matcher.py:13]: Maybe a singleton?

File: python-threatexchange/threatexchange/matcher/matcher.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            for i in input:
+                results.extend(self._match_impl(i, s_type, index))
+        return results
+
+    def _match_impl(
+        self, i: T, s_type: t.Type[SignalType], index: SignalTypeIndex
+    ) -> t.Sequence[IndexMatch]:
+        raise NotImplementedError
+
+    def _get_indicies(self) -> t.List[t.Tuple[t.Type[SignalType], SignalTypeIndex]]:
```
Comment (dougneal) [inline python-threatexchange/threatexchange/matcher/matcher.py:46]: s/indicies/indices/

**Comment by dougneal** [comment]
Nice, this is pretty much what I had in mind also.

**Pull request merged**